server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
             }
         }
     }
+
+    private static class ExtractFromSource extends IndexRouting {
+        private final String indexName;
+        private final FilterPath[] include;
+
+        ExtractFromSource(int routingNumShards, int routingFactor, String indexName, List<String> routingPaths) {
+            super(routingNumShards, routingFactor);
+            this.indexName = indexName;
+            this.include = FilterPath.compile(Set.copyOf(routingPaths));
+        }
+
+        @Override
+        public int indexShard(
+            boolean idIsAutoGenerated,
+            String id,
+            @Nullable String routing,
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
+            if (routing != null) {
+                throw new IllegalArgumentException(error("indexing with a specified routing"));
```

Comment:
I think we would want a GET request to use routing. That is obviously irrelevant for this method. But I wonder if we would then not also support an indexing request with the "right" routing value. That would allow updates and deletes too, meaning update-by-query and delete-by-query would work?

So that makes me think whether we should just calculate a routing value rather than do the hashing to shard-id here?

Reply:
The old PR calculated a routing value and that introduces a ton of complexity around what happens with it in the translog. We didn't like the idea of writing this generated routing value to the translog because its just duplicate bytes. It felt much more natural not to have to look at `routing` at all for tsdb and make the change we needed here. At least it did to me.

For what it's worth tsdb doesn't plan to support get and update and delete at all because we don't plan to have IDs at all. We have them now. And we'll probably have them in the first release. But we'd like to remove them.

This is one of the reasons why I wanted to tie this routing generation to the tsdb project - what you are proposing sound much more complex than what I did and tsdb doesn't want it.

Reply:
@henningandersen and I talked for a while about this point. We *think* it should be possible to implement stuff like `GET` and `UPDATE` based on this approach. Like I said, TSDB doesn't need them now. But its nice that we think its possible to get them from this one day if we need them badly enough.

The short version is that if we decide to use `_tsid:@timestamp` as the `_id` we can decode the routing from it and use that for `GET` and probably even `UPDATE` and update-by-query and delete-by-query.

We talked about an alternate approach that builds a `_routing` value from this information rather than a hash. Either we build it on the fly all the time or we build it and store it. Storing it is not *terrible* because it'll compress well with `_source` in stored fields. But we'd pay full price for it in the translog and over the wire. Generating on the fly is fairly wasteful compared to generating the hash. It's not a terrible idea, but it we think it'd be more complex than generating an `_id` on the fly. So that's one way we could get UPDATE and friends from this: generate the `_id` as `_tsid:@timestamp`. We're not sure we want to do that. But just having *a* way to do all this validates this approach.

Reply:
We discussed this on another channel and we think a better approach than storing the routing is to generate a `_id` on search and then extract the routing information necessary on get, index, update, delete. While there is a lot to be untangled to get there (and we may choose never to do so), it does mean that a door is open for supporting those requests. Which in turn means that the approach taken in this PR looks good to me.